src/util/cuda_util.cu

```diff
@@ -26,5 +30,21 @@ namespace micm
         ThrowInternalError(MicmInternalErrc::Cublas, file, line, msg.c_str());
       }
     }
+
+    cublasHandle_t& GetCublasHandle()
+    {
+      static std::map<int, cublasHandle_t> cublas_handles_map;
+      static std::mutex mutex;
+      int device_id;
+      CHECK_CUDA_ERROR(cudaGetDevice(&device_id), "Failed to get device ID...");
+      std::lock_guard<std::mutex> lock(mutex); // lock the mutex and generate a new cublas handle below
+      if (auto search = cublas_handles_map.find(device_id); search == cublas_handles_map.end())
+      {
+        cublasHandle_t handle;
+        CHECK_CUBLAS_ERROR(cublasCreate(&handle), "CUBLAS initialization failed..."); // create the cublas handle
+        cublas_handles_map[device_id] = handle;                                       // save the cublas handle
+      }
+      return cublas_handles_map[device_id];
+    }
```

Comment:
We don't seem to destroy the handle this way

```suggestion
    auto cublas_handle_deleter = [](cublasHandle_t* handle) {
        if (handle != nullptr) {
            CHECK_CUBLAS_ERROR(cublasDestroy(*handle), "CUBLAS finalization failed");
            delete handle;
        }
    };

    // Define the smart pointer type using decltype for the custom deleter
    using CublasHandlePtr = std::unique_ptr<cublasHandle_t, decltype(cublas_handle_deleter)>;
    
    CublasHandlePtr CreateCublasHandle() {
      cublasHandle_t handle;
      CHECK_CUBLAS_ERROR(cublasCreate(&handle), "CUBLAS initialization failed...");
      return CublasHandlePtr(new cublasHandle_t(handle), &cublasDestroy);
    }
    
    cublasHandle_t& GetCublasHandle()
    {
      static std::map<int, CublasHandlePtr> cublas_handles_map;
      static std::mutex mutex;
      int device_id;
      CHECK_CUDA_ERROR(cudaGetDevice(&device_id), "Failed to get device ID...");
      std::lock_guard<std::mutex> lock(mutex); // lock the mutex and generate a new cublas handle below
      if (auto search = cublas_handles_map.find(device_id); search == cublas_handles_map.end())
      {
        cublas_handles_map[device_id] = CreateCublasHandle();
      }
      return *cublas_handles_map[device_id];
    }
```

This uses a smart pointer with a custom deleter so that we still release the cuda memory

Reply:
Thanks @K20shores for your suggestions. I have pushed the suggested changes with some issues fixed by Copilot (hopefully they make sense to you).

One thing I want to clarify: the cublas handle itself does not seem to allocate any memory on the device (the GPU memory check passes for my implementation).

Reply:
The [cuda documentation](https://docs.nvidia.com/cuda/cublas/index.html?highlight=cublasDestroy#cublasdestroy) says that some hardware resources are released. I imagine it doesn't really matter for us since we only require the resource once and not multiple times during a single run. But it seems like it's a good idea to ensure we release them

Reply:
Agreed!